url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/ccproxy/AgentConnection.java

```diff
@@ -0,0 +1,1042 @@
+				return;
+			}
+
+			sChannel.configureBlocking(false);
+			sChannel.register(selector, SelectionKey.OP_READ);
+
+			
+			/* jesli to byl tylko dump struktury, to wychodzimy po cichu*/
+			if(this.agentName.equals("CMD")){
+				out.close();
+				in.close();
+				return;
+			}
+			idConnection += "." + agentNumber;
+
+			Iterator<String> i = queuePenalty.keySet().iterator();
+			while (i.hasNext()) {
+				String key = i.next();
+				p.log.info("key:" + key + ", value: " + queuePenalty.get(key));
+			}
+
+			if ((agentDel = p.agenci.getAgentConnectionByNumer(agentNumber)) != null) {
+				p.log.info("-ERR Mam juz agenta " + agentNumber);
+				agentDel.stopPingThread();
+				agentDel.out.println("+INFO Type 'connection_replaced', Message 'Agent "+ agentNumber+" zalogował się powtórnie. Nr. tel:"+agentExten+" IP: "+ip+"',  Ip '" +agentDel.ip+"'");
+				p.log.info("+INFO Type 'connection_replaced', Message 'Agent "+ agentNumber+" zalogował się powtórnie. Nr. tel:"+agentExten+" IP: "+ip+"',  Ip '" +agentDel.ip+"'");
+				p.asteriskQueueList.removeAgentConnection(agentDel);
+				
+				p.agenci.remove(agentDel);
+				agentDel.out.close();
+			}
+			if ((agentDel = p.agenci.getAgentConnectionByExten(agentExten)) != null) {
+				p.log.info("-ERR Mam juz agenta na tym telefonie");
+				out.println("+INFO Type 'connection_replaced', Message 'Na nr. tel: "+agentDel.agentExten+" zalogował się już agent "+ agentDel.agentNumber+" IP: "+agentDel.ip+"',  Ip '" +agentDel.ip+"'");
+				p.log.info("+INFO Type 'connection_replaced', Message 'Na nr. tel: "+agentDel.agentExten+" zalogował się już agent "+ agentDel.agentNumber+" IP: "+agentDel.ip+"',  Ip '" +agentDel.ip+"'");
+				out.close();
+				in.close();
+				return;
+			}
+			p.log.info("+INFO Licencja " + p.c.licensedAgents+", zalogowanych "+p.coreState.getLoggedAgentList().size());
+			synchronized (p.agenci) {
+			if (new Integer(p.c.licensedAgents).intValue() <= p.agenci.size()) {
+				p.log.info("-ERR Przekroczona licencja");
+				p.log.info("+INFO Type 'license_error', Message 'Brak wolnych licencji (dostępnych: " + p.c.licensedAgents+", w użyciu: "+p.coreState.getLoggedAgentList().size()+")'");
+				out.println("+INFO Type 'license_error', Message 'Brak wolnych licencji (dostępnych: " + p.c.licensedAgents+", w użyciu: "+p.coreState.getLoggedAgentList().size()+")'");
+				out.close();
+				in.close();
+				return;
+			}			
+			}
+
+			if (supervisor) {
+				out.println("+INFO Name '" + agentName + "', Extension '" + agentExten + "', AgentNumber '" + agentNumber + "', Type 'supervisor'");
+			} else {
+				out.println("+INFO Name '" + agentName + "', Extension '" + agentExten + "', AgentNumber '" + agentNumber + "', Type 'agent'");
+			}
+
+
+            if (offline){
+                out.close();
+                in.close();
+                out=new NullChannelWriter();
+            }
+
+			//wylistowanie rodzaju pauz dostępnych w systemie
+			printPauseTypesToAgent(out);
+
+			synchronized (p.agenci) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object p.agenci synchronized on line 315 should not be re-assigned in the same synchronized block on line 316. After this re-assignment, you lose the reference to the synchronized object.